torch/csrc/jit/passes/peephole_non_tensor.cpp

```diff
@@ -7,6 +7,131 @@
 namespace torch {
 namespace jit {
 
+namespace {
+
+c10::optional<IntAttr::ValueType> getConstantInt(Node& node, size_t pos) {
+  auto n = node.inputs().at(pos)->node();
+  if (n->kind() == prim::Constant &&
+      n->kindOf(attr::value) == AttributeKind::i) {
+    return n->i(attr::value);
+  }
+
+  return {};
+}
+
+c10::optional<IntAttr::ValueType> normalizeArithNode(Node& node) {
+  if (node.inputs().size() != 2) {
+    return {};
+  }
+
+  if (node.kind() == aten::mul || node.kind() == aten::add) {
+    if (auto i = getConstantInt(node, 0)) {
+      node.permuteInputs({1, 0});
+      return i;
+    }
+  }
+
+  return getConstantInt(node, 1);
+}
+
+bool tryReduceArith(Node& node);
+
+/**
+ * Remove a mul/floordiv node if it is multiplication or division by 1.
+ *
+ * @pre node is either aten::mul or aten::floordiv
+ * @return true only if the input IR node is successfully replaced.
+ */
+bool tryReduceMulAndDiv(Node& node) {
+  auto constant = normalizeArithNode(node);
+  if (!constant || *constant != 1) {
+    return false;
+  }
+
+  const auto uses = node.output()->uses();
+  for (const auto& u : uses) {
+    if (tryReduceArith(*u.user)) {
+      u.user->destroy();
+    }
+  }
+
+  node.output()->replaceAllUsesWith(node.inputs().at(0));
+  return true;
+}
+
+/**
+ * Merge an add/sub node with its users. If there exists a mul/floordiv node by
+ * 1, we firstly run tryReduceMulAndDiv() to bring all subsequent IR nodes as
+ * direct IR user node. If all updated user nodes are add/sub node with
+ * constants, we will merge the constant parts together, and replace all uses of
+ * the input node.
+ *
+ * @pre node is either aten::add or aten::sub
+ * @return true only if the input IR node is successfully replaced with its
+ * parent.
+ */
+bool tryReduceAddAndSub(Node& node) {
+  auto constant = normalizeArithNode(node);
+  if (!constant) {
+    return false;
+  }
+
+  if (*constant == 0) {
+    node.output()->replaceAllUsesWith(node.inputs().at(0));
+    return true;
+  }
+
+  std::vector<Node*> mulAndDivToRemove;
```

Comment:
This is another case where we are iterating forward looking through uses. As far as i understand, whenever we are looking at a node we are looking at all of its uses for cleanup, and then those nodes are looking in their uses for clean up, etc. 

If you have 

`add -> sub -> mul -> div`

add will recurse into sub->mul->div, then sub will recurse into mul->div, and mul will recurse in div. If we have a chain of n uses, and each use we iterate into future uses, we're doing O(n^2) work. 

I think it would be simpler and more efficient to just look at a node and see if it can be normalized wrt/ its inputs. When we look at sub, check if it can be normalized with add, when we look at mul we check if it can be normalized with sub, etc.